Content.Shared/GameObjects/Components/Chemistry/SharedSolutionContainerComponent.cs

```diff
@@ -1,115 +1,261 @@
+            UpdateAppearance();
+            Dirty();
+        }
 
-        public abstract bool TryAddSolution(Solution solution, bool skipReactionCheck = false, bool skipColor = false);
+        private void ChemicalsRemoved()
+        {
+            SolutionChanged();
+            UpdateAppearance();
+            Dirty();
+        }
 
-        public abstract bool TryRemoveReagent(string reagentId, ReagentUnit quantity);
+        private void SolutionChanged()
+        {
+            EntitySystem.Get<ChemistrySystem>()
+                .HandleSolutionChange(Owner);
+        }
 
-        /// <inheritdoc />
-        public override ComponentState GetComponentState()
+        private void ProcessReactions()
         {
-            return new SolutionContainerComponentState(Solution);
+            if (!CanReact)
+                return;
+
+            EntitySystem.Get<ChemicalReactionSystem>()
+                .FullyReactSolution(Solution, Owner, MaxVolume);
         }
 
-        public override void HandleComponentState(ComponentState? curState, ComponentState? nextState)
+        void IExamine.Examine(FormattedMessage message, bool inDetailsRange)
         {
-            base.HandleComponentState(curState, nextState);
+            if (!CanExamineContents)
+                return;
 
-            if (curState is not SolutionContainerComponentState state)
+            var prototypeManager = IoCManager.Resolve<IPrototypeManager>();
+
+            if (ReagentList.Count == 0)
             {
+                message.AddText(Loc.GetString("Contains no chemicals."));
                 return;
             }
 
-            _solution = state.Solution;
+            var primaryReagent = Solution.GetPrimaryReagentId();
+            if (!prototypeManager.TryIndex(primaryReagent, out ReagentPrototype proto))
+            {
+                Logger.Error($"{nameof(SharedSolutionContainerComponent)} could not find the prototype associated with {primaryReagent}.");
+                return;
+            }
+
+            var colorHex = Color.ToHexNoAlpha(); //TODO: If the chem has a dark color, the examine text becomes black on a black background, which is unreadable.
+            var messageString = "It contains a [color={0}]{1}[/color] " + (ReagentList.Count == 1 ? "chemical." : "mixture of chemicals.");
+
+            message.AddMarkup(Loc.GetString(messageString, colorHex, Loc.GetString(proto.PhysicalDescription)));
+        }
+
+        private void UpdateAppearance()
+        {
+            if (!Owner.TryGetComponent<SharedAppearanceComponent>(out var appearance))
+                return;
+
+            appearance.SetData(SolutionContainerVisuals.VisualState, GetVisualState());
+        }
+
+        private SolutionContainerVisualState GetVisualState()
+        {
+            var filledVolumeFraction = CurrentVolume.Float() / MaxVolume.Float();
+
+            return new SolutionContainerVisualState(Color, (byte) filledVolumeFraction);
```

Comment:
Should be multiplied by 255 I think.

Reply:
Oh, I misunderstood the point of the conversion. It now correctly converts from the 0-1 float to a 0-255 byte.